File: PrefixSum/2D PrefixSum/3548. Equal Sum Grid Partition II/solution.py
from collections import Counter

class Solution:
    def canPartitionGrid(self, grid: list[list[int]]) -> bool:
        m, n = len(grid), len(grid[0])
        total_sum = 0
        total_counts = Counter()
        
        for r in range(m):
            for c in range(n):
                val = grid[r][c]
                total_sum += val
                total_counts[val] += 1

        def check_valid(s1, s2, rows, cols, f1, ends, is_section_two=False):
            """
            s1: Sum of the section we are testing for a discount
            s2: Sum of the other section
            f1: Frequency map of Section 1 (Top or Left)
            ends: The values at the boundaries of the current section
            is_section_two: If True, we calculate Section 2's freq using (Total - f1)
            """
            if s1 == s2:
                return True
            
            target = s1 - s2
            if target <= 0:
                return False
            
            # Find if target exists in the specific section
            if not is_section_two:
                exists = f1.get(target, 0) > 0
            else:
                exists = (total_counts.get(target, 0) - f1.get(target, 0)) > 0
                
            if not exists:
                return False
            
            # Connectivity Logic:
            # If the section is a 2D block, any cell removal is safe.
            if rows > 1 and cols > 1:
                return True
            
            # If the section is 1D, target must be one of the two ends.
            return target in ends

        # --- 1. Horizontal Cuts ---
        top_sum = 0
        top_counts = Counter()
        for r in range(m - 1):
            # Update top section with the new row
            for c in range(n):
                val = grid[r][c]
                top_sum += val
                top_counts[val] += 1
            
            bottom_sum = total_sum - top_sum
            
            # Check Top Section (Ends are corners of top rectangle)
            if check_valid(top_sum, bottom_sum, r + 1, n, top_counts, 
                           {grid[0][0], grid[0][n-1], grid[r][0], grid[r][n-1]}):
                return True
            
            # Check Bottom Section (Ends are corners of bottom rectangle)
            if check_valid(bottom_sum, top_sum, m - 1 - r, n, top_counts, 
                           {grid[r+1][0], grid[r+1][n-1], grid[m-1][0], grid[m-1][n-1]}, True):
                return True

        # --- 2. Vertical Cuts ---
        left_sum = 0
        left_counts = Counter()
        for c in range(n - 1):
            # Update left section with the new column
            for r in range(m):
                val = grid[r][c]
                left_sum += val
                left_counts[val] += 1
            
            right_sum = total_sum - left_sum
            
            # Check Left Section
            if check_valid(left_sum, right_sum, m, c + 1, left_counts, 
                           {grid[0][0], grid[m-1][0], grid[0][c], grid[m-1][c]}):
                return True
            
            # Check Right Section
            if check_valid(right_sum, left_sum, m, n - 1 - c, left_counts, 
                           {grid[0][c+1], grid[m-1][c+1], grid[0][n-1], grid[m-1][n-1]}, True):
                return True

        return False

File: PrefixSum/2D PrefixSum/3548. Equal Sum Grid Partition II/solution.cpp
#include <vector>
#include <unordered_map>

using namespace std;

using LL = long long;

class Solution {
public:
    bool canPartitionGrid(vector<vector<int>>& grid) {
        int m = grid.size(), n = grid[0].size();
        LL totalSum = 0;
        unordered_map<int, int> totalFreq;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                totalSum += grid[i][j];
                totalFreq[grid[i][j]]++;
            }
        }

        // --- 1. Horizontal Cuts ---
        LL topSum = 0;
        unordered_map<int, int> topFreq;
        for (int i = 0; i < m - 1; i++) {
            for (int j = 0; j < n; j++) {
                topSum += grid[i][j];
                topFreq[grid[i][j]]++;
            }
            LL bottomSum = totalSum - topSum;
            
            // Check Top Section
            if (check(topSum, bottomSum, i + 1, n, topFreq, {grid[i][0], grid[i][n-1], grid[0][0], grid[0][n-1]})) return true;
            // Check Bottom Section (Bottom Freq = Total Freq - Top Freq)
            if (check(bottomSum, topSum, m - 1 - i, n, topFreq, {grid[i+1][0], grid[i+1][n-1], grid[m-1][0], grid[m-1][n-1]}, &totalFreq)) return true;
        }

        // --- 2. Vertical Cuts ---
        LL leftSum = 0;
        unordered_map<int, int> leftFreq;
        for (int j = 0; j < n - 1; j++) {
            for (int i = 0; i < m; i++) {
                leftSum += grid[i][j];
                leftFreq[grid[i][j]]++;
            }
            LL rightSum = totalSum - leftSum;

            // Check Left Section
            if (check(leftSum, rightSum, m, j + 1, leftFreq, {grid[0][j], grid[m-1][j], grid[0][0], grid[m-1][0]})) return true;
            // Check Right Section
            if (check(rightSum, leftSum, m, n - 1 - j, leftFreq, {grid[0][j+1], grid[m-1][j+1], grid[0][n-1], grid[m-1][n-1]}, &totalFreq)) return true;
        }

        return false;
    }

private:
    bool check(LL s1, LL s2, int rows, int cols, unordered_map<int, int>& f1, vector<int> ends, unordered_map<int, int>* totalF = nullptr) {
        if (s1 == s2) return true;
        LL target = s1 - s2;
        if (target <= 0 || target > 1e9) return false;
        int t = (int)target;

        // Determine if target exists in this section
        bool exists = false;
        if (!totalF) { // We are checking the "current" built section (Top/Left)
            exists = f1.count(t) && f1[t] > 0;
        } else { // We are checking the "remaining" section (Bottom/Right)
            exists = ((*totalF).count(t) ? (*totalF)[t] : 0) - (f1.count(t) ? f1[t] : 0) > 0;
        }

        if (!exists) return false;

        // Connectivity Rule:
        // If 2D (rows > 1 && cols > 1), any existing cell is a valid discount.
        if (rows > 1 && cols > 1) return true;
        
        // If 1D, target must be one of the ends
        for (int e : ends) if (e == t) return true;
        
        return false;
    }
};